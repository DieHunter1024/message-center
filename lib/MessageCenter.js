"use strict";

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/************************************************
 * Created By nanyuantingfeng On 5/30/16 02:44.
 ************************************************/
var MessageCenter = function () {
  function MessageCenter() {
    var maxListeners = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
    var localConsole = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : console;

    _classCallCheck(this, MessageCenter);

    var self = this;

    self._callbacks = {};
    self._events = {};
    self._console = localConsole;
    self._maxListeners = maxListeners === null ? null : parseInt(maxListeners, 10);

    return this;
  }

  _createClass(MessageCenter, [{
    key: "_addCallback",
    value: function _addCallback(eventName, callback, context, weight) {
      this._getCallbacks(eventName).push({ callback: callback, context: context, weight: weight });
      this._getCallbacks(eventName).sort(function (a, b) {
        return b.weight - a.weight;
      });
      return this;
    }
  }, {
    key: "_getCallbacks",
    value: function _getCallbacks(eventName) {
      return this._callbacks[eventName];
    }
  }, {
    key: "_getCallbackIndex",
    value: function _getCallbackIndex(eventName, callback) {
      return this._has(eventName) ? this._getCallbacks(eventName).findIndex(function (element) {
        return element.callback === callback;
      }) : -1;
    }
  }, {
    key: "_achieveMaxListener",
    value: function _achieveMaxListener(eventName) {
      var self = this;
      return self._maxListeners !== null && self._maxListeners <= this.listenersNumber(eventName);
    }
  }, {
    key: "_callbackIsExists",
    value: function _callbackIsExists(eventName, callback, context) {
      var callbackInd = this._getCallbackIndex(eventName, callback);
      var activeCallback = callbackInd !== -1 ? this._getCallbacks(eventName)[callbackInd] : void 0;

      return callbackInd !== -1 && activeCallback && activeCallback.context === context;
    }
  }, {
    key: "_has",
    value: function _has(eventName) {
      return !!this._events[eventName];
    }
  }, {
    key: "_on",
    value: function _on(eventName, callback) {
      var context = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
      var weight = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 1;

      var self = this;

      if (typeof callback !== 'function') {
        throw new TypeError(callback + " is not a function");
      }

      // If event wasn't added before - just add it
      // and define callbacks as an empty object.
      if (!this._has(eventName)) {
        self._events[eventName] = eventName;
        self._callbacks[eventName] = [];
      } else {
        // Check if we reached maximum number of listeners.
        if (this._achieveMaxListener(eventName)) {
          self._console.warn("Max listeners (" + self._maxListeners + ") for event \"" + eventName + "\" is reached!");
        }

        // Check if the same callback has already added.
        if (this._callbackIsExists.apply(this, arguments)) {
          self._console.warn("Event \"" + eventName + "\" already has the callback " + callback + ".");
        }
      }

      this._addCallback.apply(this, arguments);
      return this;
    }
  }, {
    key: "_off",
    value: function _off(eventName) {
      var callback = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;

      var self = this;
      var callbackInd = void 0;
      if (this._has(eventName)) {
        if (callback === null) {
          self._events[eventName] = undefined;
          delete self._events[eventName];
          self._callbacks[eventName] = null;
        } else {
          callbackInd = this._getCallbackIndex(eventName, callback);
          if (callbackInd !== -1) {
            this._getCallbacks(eventName).splice(callbackInd, 1);
            this._off.apply(this, arguments);
          }
        }
      }
      return this;
    }
  }, {
    key: "_emit",
    value: function _emit(eventName) {
      var custom = this._callbacks[eventName];
      var i = custom ? custom.length : 0;
      var len = arguments.length;
      var args = void 0;
      var current = void 0;

      if (i > 0 && len > 1) {
        args = [].slice.call(arguments, 1);
      }

      while (i--) {
        current = custom[i];
        if (arguments.length > 1) {
          current.callback.apply(current.context, args);
        } else {
          current.callback.apply(current.context);
        }
      }
      args = null;
      return this;
    }
  }, {
    key: "on",
    value: function on() {
      var _this = this;

      for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      var eventName = args.shift();
      eventName.split("|").forEach(function (e) {
        var _on2;

        return e && (_on2 = _this._on).call.apply(_on2, [_this, e].concat(args));
      });
      return this;
    }
  }, {
    key: "once",
    value: function once(eventName, callback) {
      var _this2 = this;

      var context = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
      var weight = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 1;

      var onceCallback = function onceCallback() {
        for (var _len2 = arguments.length, args = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
          args[_key2] = arguments[_key2];
        }

        _this2.off(eventName, onceCallback);
        return callback.apply(context, args);
      };
      return this.on(eventName, onceCallback, context, weight);
    }
  }, {
    key: "off",
    value: function off() {
      var _this3 = this;

      for (var _len3 = arguments.length, args = Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
        args[_key3] = arguments[_key3];
      }

      var eventName = args.shift();
      eventName.split("|").forEach(function (e) {
        var _off2;

        return e && (_off2 = _this3._off).call.apply(_off2, [_this3, e].concat(args));
      });
      return this;
    }
  }, {
    key: "emit",
    value: function emit() {
      var _this4 = this;

      for (var _len4 = arguments.length, args = Array(_len4), _key4 = 0; _key4 < _len4; _key4++) {
        args[_key4] = arguments[_key4];
      }

      var eventName = args.shift();
      eventName.split("|").forEach(function (e) {
        var _emit2;

        return e && (_emit2 = _this4._emit).call.apply(_emit2, [_this4, e].concat(args));
      });
      return this;
    }
  }, {
    key: "fire",
    value: function fire() {
      return this.emit.apply(this, arguments);
    }
  }, {
    key: "un",
    value: function un() {
      return this.off.apply(this, arguments);
    }
  }, {
    key: "clear",
    value: function clear() {
      var self = this;
      self._events = {};
      self._callbacks = {};
      return this;
    }
  }, {
    key: "listenersNumber",
    value: function listenersNumber(eventName) {
      return this._has(eventName) ? this._callbacks[eventName].length : null;
    }
  }]);

  return MessageCenter;
}();

module.exports = MessageCenter;