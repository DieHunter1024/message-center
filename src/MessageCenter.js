/************************************************
 * Created By nanyuantingfeng On 5/30/16 02:44.
 ************************************************/
class MessageCenter {

  constructor(maxListeners = null, localConsole = console) {
    const self = this

    self._callbacks = {};
    self._events = {}
    self._console = localConsole;
    self._maxListeners = maxListeners === null ? null : parseInt(maxListeners, 10);

    return this;
  }

  _addCallback(eventName, callback, context, weight) {
    this._getCallbacks(eventName).push({ callback, context, weight });
    this._getCallbacks(eventName).sort((a, b) => b.weight - a.weight);
    return this;
  }

  _getCallbacks(eventName) {
    return this._callbacks[eventName];
  }

  _getCallbackIndex(eventName, callback) {
    return this._has(eventName) ?
      this._getCallbacks(eventName)
        .findIndex(element => element.callback === callback) : -1;
  }

  _achieveMaxListener(eventName) {
    const self = this;
    return (self._maxListeners !== null && self._maxListeners <= this.listenersNumber(eventName));
  }

  _callbackIsExists(eventName, callback, context) {
    const callbackInd = this._getCallbackIndex(eventName, callback);
    const activeCallback = callbackInd !== -1 ?
      this._getCallbacks(eventName)[callbackInd] : void 0;

    return (callbackInd !== -1 && activeCallback && activeCallback.context === context);
  }

  _has(eventName) {
    return !!this._events[eventName]
  }

  _on(eventName, callback, context = null, weight = 1) {
    const self = this;

    if (typeof callback !== 'function') {
      throw new TypeError(`${callback} is not a function`);
    }

    // If event wasn't added before - just add it
    // and define callbacks as an empty object.
    if (!this._has(eventName)) {
      self._events[eventName] = eventName;
      self._callbacks[eventName] = [];
    } else {
      // Check if we reached maximum number of listeners.
      if (this._achieveMaxListener(eventName)) {
        self._console.warn(`Max listeners (${self._maxListeners}) for event "${eventName}" is reached!`);
      }

      // Check if the same callback has already added.
      if (this._callbackIsExists(...arguments)) {
        self._console.warn(`Event "${eventName}" already has the callback ${callback}.`);
      }
    }

    this._addCallback(...arguments);
    return this;
  }

  _off(eventName, callback = null) {
    const self = this;
    let callbackInd;
    if (this._has(eventName)) {
      if (callback === null) {
        self._events[eventName] = undefined;
        delete self._events[eventName]
        self._callbacks[eventName] = null;
      } else {
        callbackInd = this._getCallbackIndex(eventName, callback);
        if (callbackInd !== -1) {
          this._getCallbacks(eventName).splice(callbackInd, 1);
          this._off(...arguments);
        }
      }
    }
    return this;
  }

  _emit(eventName) {
    const custom = this._callbacks[eventName];
    let i = custom ? custom.length : 0;
    let len = arguments.length;
    let args;
    let current;

    if (i > 0 && len > 1) {
      args = [].slice.call(arguments, 1)
    }

    while (i--) {
      current = custom[i];
      if (arguments.length > 1) {
        current.callback.apply(current.context, args);
      } else {
        current.callback.apply(current.context);
      }
    }
    args = null;
    return this;
  }

  on(...args) {
    let eventName = args.shift();
    eventName.split("|").forEach(e => e && this._on.call(this, e, ...args));
    return this;
  }

  once(eventName, callback, context = null, weight = 1) {
    const onceCallback = (...args) => {
      this.off(eventName, onceCallback);
      return callback.apply(context, args);
    };
    return this.on(eventName, onceCallback, context, weight);
  }

  off(...args) {
    let eventName = args.shift();
    eventName.split("|").forEach(e => e && this._off.call(this, e, ...args));
    return this;
  }

  emit(...args) {
    let eventName = args.shift();
    eventName.split("|").forEach(e => e && this._emit.call(this, e, ...args));
    return this;
  }

  fire() {
    return this.emit.apply(this, arguments)
  }

  un() {
    return this.off.apply(this, arguments);
  }

  clear() {
    const self = this;
    self._events = {};
    self._callbacks = {};
    return this;
  }

  listenersNumber(eventName) {
    return this._has(eventName) ? this._callbacks[eventName].length : null;
  }
}

module.exports = MessageCenter;
