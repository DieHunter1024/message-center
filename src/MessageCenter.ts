/**************************************************
 * Created by nanyuantingfeng on 20/12/2016 14:51.
 **************************************************/

interface IDefer {
  promise: Promise<any>;
  resolve: Function;
  reject: Function;
}

function Defer(): IDefer {
  let resolve;
  let reject;

  const promise = new Promise((a, b) => {
    resolve = a;
    reject = b;
  });

  return {promise, resolve, reject};
}

function prefixEventName(name): string {
  return `@@A0F2F71915C05BE72D17F48B2A49CEAD:${name}`;
}

export default class MessageCenter {

  private __handlers = {};

  private __events = {};

  private readonly __maxListeners = null;

  private __watchHandlersMap = new WeakMap();

  private __addHandler(name, handler, context, weight): this {
    this.__getHandlers(name).push({handler, context, weight});
    this.__getHandlers(name).sort((a, b) => b.weight - a.weight);
    return this;
  }

  private __getHandlers(name) {
    return this.__handlers[name];
  }

  private __getHandlerIndex(name, handler): number {
    return this.has(name)
      ? this.__getHandlers(name).findIndex(element => element.handler === handler)
      : -1;
  }

  private __achieveMaxListener(name): boolean {
    return (this.__maxListeners !== null && this.__maxListeners <= this.listenersLength(name));
  }

  private __handlerIsExists(name, handler, context): boolean {
    const handlerInd = this.__getHandlerIndex(name, handler);
    const activeHandler = handlerInd !== -1 ? this.__getHandlers(name)[handlerInd] : void 0;
    return (handlerInd !== -1 && activeHandler && activeHandler.context === context);
  }

  private __on(name: string, handler: Function, context?: any, weight = 1): this {

    if (typeof handler !== 'function') {
      throw new TypeError(`${handler} is not a function`);
    }

    if (!this.has(name)) {
      this.__events[name] = name;
      this.__handlers[name] = [];
    } else {
      // Check if we reached maximum number of listeners.
      if (this.__achieveMaxListener(name)) {
        console.warn(`Max listeners (${this.__maxListeners}) for event "${name}" is reached!`);
      }

      // Check if the same handler has already added.
      if (this.__handlerIsExists.apply(this, arguments)) {
        console.warn(`Event "${name}" already has the handler ${handler}.`);
      }
    }

    this.__addHandler.apply(this, arguments);
    return this;
  }

  private __un(name: string, handler = null) {
    let handlerInd;
    if (this.has(name)) {
      if (handler === null) {
        this.__events[name] = undefined;
        delete this.__events[name];
        this.__handlers[name] = null;
      } else {
        handler = this.__getHandlerInMap(handler);
        handlerInd = this.__getHandlerIndex(name, handler);
        if (handlerInd !== -1) {
          this.__getHandlers(name).splice(handlerInd, 1);
          this.__un.apply(this, arguments);
        }
      }
    }
    return this;
  }

  private __emit(name: string, ...args): this {
    const custom = this.__handlers[name];
    let i = custom ? custom.length : 0;
    let current;

    while (i--) {
      current = custom[i];
      current.handler.apply(current.context, args);
    }

    return this;
  }

  private __setHandlerInMap(handler, realHandler) {
    this.__watchHandlersMap.set(handler, realHandler);
  }

  private __getHandlerInMap(handler) {
    return this.__watchHandlersMap.get(handler) || handler;
  }

  constructor(maxListeners ?: number|string|null) {
    this.__maxListeners = maxListeners === null ? null : parseInt(String(maxListeners), 10);
  }

  has(name: string): boolean {
    return !!this.__events[name];
  }

  on(name: string, handler: Function, context?: any, weight = 1): this {
    name.split('|').forEach(e => e && this.__on(e, handler, context, weight));
    return this;
  }

  /*********************************************
   * 单次监听
   */
  once(name: string, handler: Function, context?: any, weight = 1): this {
    const fn = (...args) => {
      this.un(name, fn);
      return handler.apply(context, args);
    };
    return this.on(name, fn, context, weight);
  }

  /*********************************************
   * 解除监听
   */
  un(name: string, handler ?: Function): this {
    name.split('|').forEach(e => e && this.__un(e, handler));
    return this;
  }

  /**********************************************
   * 触发监听
   */
  emit(name: string, ...args): this {
    name.split('|').forEach(e => e && this.__emit(e, ...args));
    return this;
  }

  /******************************************
   * 清空当前实例中所有的监听
   */
  clear(): this {
    this.__events = {};
    this.__handlers = {};
    return this;
  }

  /****************************************
   * 检测监听器的长度
   * @param name
   * @returns {number}
   */
  listenersLength(name: string): number {
    return this.has(name) ? this.__handlers[name].length : 0;
  }

  /*********************************
   * 等待结果返回
   */
  watch(name: string, handler: Function, context ?: any): this {
    const fn = (...data) => {
      this.emit(prefixEventName(name), handler(...data));
    };

    this.on(name, fn, context);
    this.__setHandlerInMap(handler, fn);
    return this;
  }

  /*************************
   * 触发结果返回
   */
  invoke(name: string, ...args): Promise<any> {
    const {promise, resolve, reject} = Defer();

    if (!this.listenersLength(name)) {
      reject(`have no watcher at event(${name})`);
    } else {
      this.once(prefixEventName(name), resolve);
      this.emit(name, ...args);
    }

    return promise;
  }
}
