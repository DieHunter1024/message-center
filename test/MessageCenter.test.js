/**************************************************
 * Created by nanyuantingfeng on 10/03/2017 16:39.
 **************************************************/
import MessageCenter from '../src/MessageCenter'

test('on/emit', () => {

  const bus = new MessageCenter()

  bus.on('aa|bb', data => {
    expect(data).toBe(999)
  })

  bus.emit('aa', 999)

  bus.emit('bb', 999)

})

test('watch/invoke', () => {
  const bus = new MessageCenter()

  let fff = (a, b, c) => {
    expect(a).toBe(1)
    expect(b).toBe(2)
    expect(c).toBe(3)
    return {a, b, c}
  }

  bus.watch('kkk', fff)

  let x = bus.invoke('kkk', 1, 2, 3)

  expect(x).toBeInstanceOf(Promise)

  x.then(d => {
    expect(d).toEqual({a: 1, b: 2, c: 3})
  })

  bus.un('kkk', fff)

  console.log(bus._handlers)

  expect(bus._handlers['kkk']).toEqual([])
})

test('noWatch/invoke1', (done) => {
  const bus = new MessageCenter()

  let x = bus.invoke('kkk', 1, 2, 3)

  x.catch(e => {
    expect(e).toBe('have no watcher at event(kkk)')
    done()
  })

})

test('noWatch/invoke2', (done) => {
  const bus = new MessageCenter()

  let fff = (a, b, c) => {
    expect(a).toBe(1)
    expect(b).toBe(2)
    expect(c).toBe(3)
    return {a, b, c}
  }

  bus.watch('kkk', fff)

  let y = bus.invoke('kkk', 1, 2, 3)

  y.then(data => {
    expect(data).toEqual({a: 1, b: 2, c: 3})
    done()
  })

})
