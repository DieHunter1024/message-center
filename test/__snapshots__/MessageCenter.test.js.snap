// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test clear 1`] = `
Object {
  "kkk": Array [
    Object {
      "context": null,
      "handler": [Function],
      "weight": 1,
    },
  ],
  "kkk2": Array [
    Object {
      "context": null,
      "handler": [Function],
      "weight": 1,
    },
  ],
  "kkk3": Array [
    Object {
      "context": null,
      "handler": [Function],
      "weight": 1,
    },
  ],
  "kkk4": Array [
    Object {
      "context": null,
      "handler": [Function],
      "weight": 1,
    },
  ],
}
`;
